### PR TITLE
everything

### PR DESCRIPTION
**Pull request by davidchambers**
We've spent a lot of time developing the infrastructure surrounding the various Sanctuary projects. Here are some of the pull requests:

  - sanctuary-js/sanctuary#36 add documentation and Transcribe-powered prepublish script
  - sanctuary-js/sanctuary#213 license: update copyright information and automate future updates
  - sanctuary-js/sanctuary#324 documentation: use version-specific project URLs
  - sanctuary-js/sanctuary#363 lint: use a Markdown parser to check link definitions
  - sanctuary-js/sanctuary#420 lint: ensure merge commit summary will not exceed 72 characters

When creating a new Sanctuary project we usually copy an existing makefile, a couple of shell scripts, and a dozen or so dev dependencies. There are two problems with this approach:

  - improving a shell script involves making the same change in multiple repositories; and
  - the same dozen or so dev dependencies must be kept up to date in multiple repositories.

To address these problems, this pull request brings together (and improves) a number of scripts currently found in various Sanctuary repositories.

These compare views provide an idea of the amount of common code which we'll be able to remove from the various Sanctuary repositories:

  - <https://github.com/sanctuary-js/sanctuary/compare/master...davidchambers/scripts>
  - <https://github.com/sanctuary-js/sanctuary-def/compare/master...davidchambers/scripts>

Note that the diff is less impressive for sanctuary-def, but this is because some of the infrastructure added to the Sanctuary repository has not yet propagated to other repositories.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,28 @@
+SHELL = /bin/bash
+
+NPM = npm
+SHELLCHECK = shellcheck
```
Comment (Avaq) [inline Makefile]: ShellCheck looks like a nice tool :)

> written in Haskell, if you're into that sort of thing

I'm sold!

File: README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# sanctuary-scripts
+
+Shell scripts used in multiple Sanctuary projects.
```
Comment (Avaq) [inline README.md]: Should there be more information here?

File: bin/generate-readme [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+set -euf -o pipefail
+
+transcribe() {
+  local -r project="$(node --print 'require("./package.json").repository.url.replace(/[.]git$/, "").split("/").slice(-2).join("/")')"
```
Comment (Avaq) [inline bin/generate-readme]: The `repository` field in a `package.json` doesn't have to be an object with a `url` property. It [can be a string representation](https://docs.npmjs.com/files/package.json#repository) with a variety of formats. Do we want to support these formats? If not, we should mention that in the README.

File: bin/generate-readme [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  local -r version="${VERSION?Environment variable not set}"
+  local -r url="https://github.com/$project/blob/v$version/{filename}#L{line}"
```
Comment (Avaq) [inline bin/generate-readme]: I see we rely explicitly on version being prefixed by `v`. I don't see a script for XYZ specified in this repo, so this means that one has to configure XYZ properly for `sanctuary-scripts` to work. Maybe we can add a `release.sh` script or something to this repo which ensures the correct tagging convention? Or if not, we should add a note to the README.

File: bin/lint-commit-message [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+pull_request_number() {
+  {
+    git log --grep '^Merge pull request #[1-9]' --pretty=format:%s \
+    | cut -d '#' -f 2 \
+    | cut -d ' ' -f 1 \
+    | sort --numeric-sort --reverse ;
+    echo 0 ;
+  } \
+  | head -n 1 \
+  | xargs expr 1 + 25 +
```
Comment (Avaq) [inline bin/lint-commit-message]: Rather than adding an arbitrary buffer to the PR *number*, can't we just add a single *character* buffer? So an additional reserved character to allow for the expansion of the PR number? That would mean that:

After the **1st** PR, we reserve **9** slots
Once we reach **10** PR's, we reserve **990** slots
Once we reach **100** PR's, we reserve **9900** slots
...etc...

To me this approach seems less arbitrary, and more scalable. On the other hand, it might be overkill, and it may be upsetting to have 90% of all merge commit messages needlessly limited to -1 characters.

File: bin/lint-commit-message [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+project_name() {
+  git config --get remote.origin.url \
+  | sed -e 's,^git@github[.]com:,,' -e 's,^https://github[.]com/,,' \
```
Comment (Avaq) [inline bin/lint-commit-message]: That's a nice way to avoid escape-hell (`\\\\.`) in this case. I will remember. :D

File: bin/generate-readme (hunk 3 above)
Comment (Avaq) [reply bin/generate-readme]: I didn't know about the `--print` option in Node. That's great!

File: bin/update-copyright-year [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+}
+substitute <"$license" >"$updated"
+
+if (( count == 0 )) ; then
+  rm "$updated"
+  echo "Copyright attribution not found for $creator" >&2
+  exit 1
+fi
+
+mv "$updated" "$license"
```
Comment (Avaq) [inline bin/update-copyright-year]: Shouldn't this be forced? Won't bash complain that the file is already at this destination?

**Review by Avaq — COMMENTED** [review]
I like this project a lot! I enjoy seeing advanced\* bash scripting.

<sub>\* By *advanced*, I mean anything I couldn't pull off myself.</sub>

File: README.md (hunk 2 above)
Comment (davidchambers) [reply README.md]: Quite possibly. What sort of information do you have in mind?

File: bin/generate-readme (hunk 3 above)
Comment (davidchambers) [reply bin/generate-readme]: My goal is to make using these scripts in Sanctuary projects (and other projects with the same conventions) as convenient as possible. This means I'm willing to make the scripts less flexible in order to avoid configuration (in some cases default values can be used to let us have our :cake: and eat it too).

We could add a `lint-package-json` script which performs some __package.json__-specific checks in addition to the sanity check performed by `lint-json`. It could assert that `repository.url` is present and that its value is suitable. What do you think of this idea?

File: bin/generate-readme (hunk 4 above)
Comment (davidchambers) [reply bin/generate-readme]: Good point. I had not noticed this unstated assumption. I had been unsure as to whether sanctuary-scripts should manage the xyz dependency; you've made it clear to me that it should.

File: bin/lint-commit-message (hunk 5 above)
Comment (davidchambers) [reply bin/lint-commit-message]: > To me this approach seems less arbitrary, and more scalable.

I dislike arbitrariness. I like your suggestion. I will make the change. :)

File: bin/lint-commit-message (hunk 6 above)
Comment (davidchambers) [reply bin/lint-commit-message]: Also, using `s,,,` rather than `s///` avoids lots of confusion when matching URLs. :)

File: bin/update-copyright-year (hunk 7 above)
Comment (davidchambers) [reply bin/update-copyright-year]: That's certainly a reasonable expectation. By default, though, `mv` will quite happily replace an existing file (there are flags for changing this behaviour).

File: bin/generate-readme (hunk 3 above)
Comment (davidchambers) [reply bin/generate-readme]: :zap:

File: bin/generate-readme (hunk 4 above)
Comment (davidchambers) [reply bin/generate-readme]: :zap:

File: bin/lint-commit-message (hunk 5 above)
Comment (davidchambers) [reply bin/lint-commit-message]: :zap:

File: README.md (hunk 2 above)
Comment (davidchambers) [reply README.md]: :zap:

File: README.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+Uses [sanctuary-style][sanctuary-style] to lint ES6 code. Accepts any number
+of arguments to be forwarded to [`eslint`][eslint].
+
+## `lint-json`
+
+Asserts that the specified JSON files exist and are neatly formatted.
+
+## `lint-package-json`
+
+Asserts that __package.json__ exists and contains certain fields with
```
Comment (Avaq) [inline README.md]: *~certain~ some important*? (to match `check-required-files` and remove air of mystery)

File: README.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+of arguments to be forwarded to [`eslint`][eslint].
+
+## `lint-es6`
+
+Uses [sanctuary-style][sanctuary-style] to lint ES6 code. Accepts any number
+of arguments to be forwarded to [`eslint`][eslint].
+
+## `lint-json`
+
+Asserts that the specified JSON files exist and are neatly formatted.
```
Comment (Avaq) [inline README.md]: It looks like the script doesn't do what this says. It should say something like: *Asserts that the specified JSON files exist and have no syntax error, and then formats these files.*

Comment (davidchambers) [reply README.md]: It really does assert that the files are neatly formatted. After reformatting a file it runs `git --no-pager diff --exit-code -- "$file"`. The `--exit-code` flag is important here:

> ```
>        --exit-code
>            Make the program exit with codes similar to diff(1).
>            That is, it exits with 1 if there were differences
>            and 0 means no differences.
```

Comment (Avaq) [reply README.md]: Ahh. I see. But it does leave the working tree "dirty" (with neat files) afterwards?

File: bin/release [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+declare -r increment="$1"
+owner="$(get repo-owner)"
+name="$(get repo-name)"
+prefix="$(get version-tag-prefix)"
+
+node_modules/.bin/xyz \
+  --increment "${increment}" \
+  --repo "git@github.com:${owner}/${name}.git" \
+  --script scripts/prepublish \
```
Comment (Avaq) [inline bin/release]: Now that there's way to configure some of these scripts, should this field remain hard-coded? Maybe

```config
  prepublish-script = scripts/prepublish
```

Allowing it to be emptied, similarly to the `version-prefix` option?

File: README.md (hunk 9 above)
Comment (davidchambers) [reply README.md]: Having put this script through its paces I no longer think involving `git` is a good idea. I'll rework the script to use `cp` and `diff` instead. This will avoid making a mess, and will allow linting of untracked JSON files.

File: bin/release (hunk 10 above)
Comment (davidchambers) [reply bin/release]: Good idea. I'll make this change. We can keep `scripts/prepublish` as the default value, and have `check-required-files` assert that the (explicitly or implicitly) specified file is present and executable.

Comment (Avaq) [reply bin/release]: :+1:

File: README.md (hunk 8 above)
Comment (davidchambers) [reply README.md]: :zap:

File: bin/release (hunk 10 above)
Comment (davidchambers) [reply bin/release]: I decided not to make this change after all, as we're already relying on convention for overriding the behaviour of scripts. To override `sanctuary-lint`, for example, one would create an executable at __scripts/lint__.

Note that this pull request includes a default prepublish script suitable for use in most Sanctuary projects.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    repo-name = sanctuary-type-classes
+    ```
+
+4.  Add a file named __.eslintrc.json__ in the project's root directory,
+    with the following JSON configuration:
+
+    ```json
+    {
+      "root": true,
+      "extends": ["./node_modules/sanctuary-style/eslint-es3.json"]
```
Comment (Avaq) [inline README.md:43]: I think you can use a short notation, like `"sanctuary-style/eslint-es3".`

File: README.md [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+| `repo-owner`                  |               | The name of the GitHub user who owns the repository.                          |
+| `repo-name`                   |               | The name of the GitHub repository.                                            |
+| `default-branch`              | `master`      | The name of the repository's default branch.                                  |
+| `min-branch-coverage`         | `100`         | The minimum acceptable branch coverage (as a percentage).                     |
+| `source-files`                | `index.js`    | Space-separated list of filenames. Globbing is supported.                     |
+| `heading-level`               | `4`           | The `<h[1-6]>` level of headings transcribed from `heading-prefix` comments.  |
+| `heading-prefix`              | `#`           | The character which follows `//` to signify a heading to transcribe.          |
+| `comment-prefix`              | `.`           | The character which follows `//` to signify documentation to transcribe.      |
+| `version-tag-prefix`          | `v`           | The prefix of annotated version tags (use _null_ for `X.Y.Z` format).         |
```
Comment (Avaq) [inline README.md]: It's not immediately clear to me what is meant by use _null_. Do I set the value to literal `null`? Or do I omit the value part of my expression ?

File: bin/test [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+if [[ "${RECURSING-false}" == false ]] && [[ -x scripts/test ]] ; then
+  RECURSING=true scripts/test
+  exit 0
+fi
+
+branches="$(get min-branch-coverage)"
+
+node_modules/.bin/istanbul cover node_modules/.bin/_mocha
+node_modules/.bin/istanbul check-coverage --branches "$branches"
```
Comment (Avaq) [inline bin/test:12]: Did you consider updating to [`nyc`](https://github.com/istanbuljs/nyc)?

File: bin/update-copyright-year [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+# shellcheck source=functions
+source "${BASH_SOURCE%/*}/../sanctuary-scripts/functions"
+
+if [[ "${RECURSING-false}" == false ]] && [[ -x scripts/update-copyright-year ]] ; then
+  RECURSING=true scripts/update-copyright-year
+  exit 0
+fi
+
+declare -r creator="${1-Sanctuary}"
+declare -r license="${2-LICENSE}"
```
Comment (Avaq) [inline bin/update-copyright-year]: What's the logic which determines a script using command line arguments over a value in `.config`? Incidentally, could you teach me about `${1-foo}`-syntax?

**Review by Avaq — COMMENTED** [review]
I noticed you squashed, so I gave it a last good look.

File: README.md (hunk 11 above)
Comment (davidchambers) [reply README.md:43]: Thanks. I didn't know that. I'll keep it as it is for consistency with the handful of projects I updated over the weekend.

File: README.md (hunk 12 above)
Comment (davidchambers) [reply README.md]: I agree that it's unclear.

I consulted the [Bash Reference Manual][1] many times in the days prior to writing this. It uses _null_, _null string_, and _null value_ to refer to the empty string (in JavaScript parlance).

I'm attempting to convey that a user should add

    version-tag-prefix =

to her __.config__ if she wants to use `X.Y.Z` format rather than `vX.Y.Z` format. The problem with referring to the empty string is it suggests she should add

    version-tag-prefix = ''

instead, which is incorrect.

Can you think of a better explanation?


[1]: https://www.gnu.org/software/bash/manual/bashref.html

File: bin/test (hunk 13 above)
Comment (davidchambers) [reply bin/test:12]: No. Do you think we should do so? What are the advantages?

File: bin/update-copyright-year (hunk 14 above)
Comment (davidchambers) [reply bin/update-copyright-year]: > What's the logic which determines a script using command line arguments over a value in `.config`?

This script is inconsistent in this respect, but for good reasons (I believe). It seems to me the script most likely to be used in non-Sanctuary projects. It's thus important that the name of the licence file is configurable (`check-required-files` actually requires a file named __LICENSE__, so this is only relevant to projects using scripts à la carte).

I'd like to support those who wish to use *only* this script. It would be burdensome and a little messy to require a configuration file in order to use a single shell script.

Perhaps I'm overthinking this. It would certainly be more consistent to move these to __.config__, and to remove from `check-required-files` the hard-coded reference to __LICENSE__. Is this what you suggest?

> Incidentally, could you teach me about `${1-foo}`-syntax?

The first thing to know is that `$var` is shorthand for `${var}`:

```console
$ VAR=hello bash -c 'var=world ; echo "$VAR $var"'
hello world

$ VAR=hello bash -c 'var=world ; echo "${VAR} ${var}"'
hello world
```

As you're no doubt aware, `$1`, `$2`, `$3`, etc. refer to command-line arguments. We now know that these can be written as `${1}`, `${2}`, `${3}`, etc.

[§ 3.5.3 Shell Parameter Expansion][1]:

> `${parameter:-word}`
>
> If _parameter_ is unset or null, the expansion of _word_ is substituted. Otherwise, the value of _parameter_ is substituted.

```console
$ bash -c 'foo=42 ; bar= ; echo "foo: ${foo:-default}, bar: ${bar:-default}, baz: ${baz:-default}"'
foo: 42, bar: default, baz: default
```

We are omitting the colon, though, which changes the handling of null:

> When not performing substring expansion, using the form described below (e.g., ‘`:-`’), Bash tests for a parameter that is unset or null. Omitting the colon results in a test only for a parameter that is unset. Put another way, if the colon is included, the operator tests for both `parameter`’s existence and that its value is not null; if the colon is omitted, the operator tests only for existence.

```console
$ bash -c 'foo=42 ; bar= ; echo "foo: ${foo-default}, bar: ${bar-default}, baz: ${baz-default}"'
foo: 42, bar: , baz: default
```

So, `${1-Sanctuary}` evaluates to the value of the first command-line argument if at least one command-line argument is provided; `Sanctuary` otherwise. As is so often the case with Bash, this convenient functionality is hidden behind arcane syntax few will ever discover.


[1]: https://www.gnu.org/software/bash/manual/bashref.html#Shell-Parameter-Expansion

File: README.md (hunk 12 above)
Comment (davidchambers) [reply README.md]: I went with this:

> The prefix of annotated version tags (use `version-tag-prefix =` for `` ``).

:zap:

File: bin/check-required-files [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+check() {
+  if test "$1" "$2" ; then
+    pass "test $1 $2"
+  else
+    fail "test $1 $2"
+    (( failures += 1 ))
+  fi
+}
+
+check -f .circleci/config.yml
```
Comment (Avaq) [inline bin/check-required-files]: It's unfortunate that this script creates a circleci lock-in. I've tried using sanctuary-script on my own project which uses Travis, and avoiding this check is not that straightforward.

Comment (davidchambers) [reply bin/check-required-files]: Good point! Shall we simply remove this check?

Comment (Avaq) [reply bin/check-required-files]: That would certainly solve my issue. Perhaps we can also check that all `source-files` are present (not sure if we want to though), and in that spirit, we could add a `dev-files` options or something that defaults to `.circleci/config.yml`. In that case I could avoid this check by adding the following to my config:

```text
dev-files = .travis.yml
```

Comment (davidchambers) [reply bin/check-required-files]: The benefit of such a check is marginal if one must remember to enable the check. In other words, one could just as easily `touch .travis.yml && git add .travis.yml` as add `dev-files = .travis.yml` to __.config__.

I've removed the check above. :)

Comment (Avaq) [reply bin/check-required-files]: :+1:

File: bin/lint-readme [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+fi
+
+VERSION=0.0.0 run generate-readme
+
+node_modules/.bin/remark \
+  --frail \
+  --no-stdout \
+  --use remark-lint-no-undefined-references \
+  --use remark-lint-no-unused-definitions \
+  --use validate-links \
```
Comment (Avaq) [inline bin/lint-readme]: The `validate-links` plugin fails all README's generated by `generate-readme`, because it does not take HTML heading elements into account when verifying links to anchors.

File: bin/test (hunk 13 above)
Comment (Avaq) [reply bin/test:12]: Let's wait for a tool that uses V8 native code coverage reporting.

File: bin/generate-readme [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+  declare -A dependencies
+  while IFS=$'\t' read -r k v ; do dependencies["$k"]="$v" ; done <<<"$(node --eval '
+    const pkg = require("./package.json");
+    const log = o => Object.keys(o).forEach(k => { console.log(k + "\t" + o[k]); });
+    log(pkg.dependencies);
+    log(pkg.devDependencies);
+  ')"
+
+  local -r pattern1='^<h4 name="([^"]*)#([^"]*)">'
+  local -r pattern2='^(.+) ([Vv]):(.+)/([^#]+)'  # V => 1.2.3 ; v => v1.2.3
```
Comment (davidchambers) [inline bin/generate-readme]: :warning: This must be updated in accordance with sanctuary-js/sanctuary#468.

File: bin/test (hunk 13 above)
Comment (Avaq) [reply bin/test:12]: It just spawned into existence https://github.com/bcoe/c8

File: bin/lint-package-json [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    pass "$1"
+  else
+    fail "$1"
+    (( failures += 1 ))
+  fi
+}
+
+assert "pkg.name === '$name' || pkg.name === '@$owner/$name'"
+
+assert "matches(pkg.version, /^[0-9]+[.][0-9]+[.][0-9]+$/)"
```
Comment (davidchambers) [inline bin/lint-package-json]: :warning: This will fail for pre-release versions. We should either remove this assertion or relax the pattern.

Comment (Avaq) [reply bin/lint-package-json]: I used `/^\d+[.]\d+[.]\d+(?:-[a-z0-9.]+)?(?:[+][a-z0-9.]+)?$/` in one of my own projects.

Comment (davidchambers) [reply bin/lint-package-json]: Let's just use `/./` here. All I really want to ensure is that the field is present.

Comment (Avaq) [reply bin/lint-package-json]: The reason I've used such specific checking, is because another part of my code relies on this value having that specific format.

The same might be said about [`bin/release`](https://github.com/sanctuary-js/sanctuary-scripts/blob/davidchambers/everything/bin/release) depending on the format of `pkg.version`, so I think it's worth considering the strict validation so to prevent `release` from failing in an unexpected way.

Comment (davidchambers) [reply bin/lint-package-json]: `release` would fail with a message such as `Invalid version: XXX` before performing any irreversible actions, so I don't think extra validation is warranted. Furthermore, one should rarely need to edit `pkg.version` manually so invalid values are unlikely.

**Review by Avaq — APPROVED** [review]

**Review by Avaq — APPROVED** [review]
Ready to merge?
